src/qrisp/interface/backend.py

```diff
@@ -0,0 +1,174 @@
+
+    Parameters
+    ----------
+    name : str or None
+        Optional user-defined name for the backend.
+        Defaults to the class name.
+
+    options : Mapping or None
+        Runtime execution options for the backend.
+        If omitted, :meth:`_default_options` is used.
+
+    """
+
+    def __init__(self, name: Optional[str] = None, options: Optional[Mapping] = None):
+        """
+        Initialize the backend.
+
+        Parameters
+        ----------
+        name : str or None
+            Optional user-defined name for this backend.
+            Defaults to the class name.
+
+        options : Mapping or None
+            Mapping of runtime execution options for the backend.
+            If omitted, the backend uses the class-level default options from
+            :meth:`_default_options`.
+
+        """
+
+        self.name = name or self.__class__.__name__
+
+        if options is None:
+            options = self._default_options()
+
+        if not isinstance(options, Mapping):
+            raise TypeError("options must be a dict-like mapping")
+
+        self._options = copy(options)
+
+    # ----------------------------------------------------------------------
+    # Abstract interface
+    # ----------------------------------------------------------------------
+
+    @abstractmethod
+    def run(self, *args, **kwargs) -> Any:
+        """
+        Execute one or more circuits/operations on the backend.
+
+        Subclasses may override this method with a more specific signature.
+
+        Returns
+        -------
+        Any
+            Backend-specific result of the execution.
+
+        """
+        raise NotImplementedError
+
+    # ----------------------------------------------------------------------
+    # Runtime options
+    # ----------------------------------------------------------------------
+
+    @classmethod
+    def _default_options(cls) -> Mapping:
+        """
+        Default runtime options for the backend.
+
+        Child classes may override this method to provide custom default options,
+        or the defaults may be overridden entirely by passing an ``options``
+        mapping to the constructor.
+        """
+        return {"shots": 1024}
+
+    @property
+    def options(self):
+        """Current runtime options."""
+        return self._options
+
+    def update_options(self, **kwargs) -> None:
+        """Update existing runtime options, rejecting unknown keys."""
+
+        for key, val in kwargs.items():
+            if key not in self._options:
+                raise AttributeError(
+                    f"'{key}' is not a valid backend option for {self.__class__.__name__}. "
+                    f"Valid options: {list(self._options.keys())}"
+                )
+            self._options[key] = val
+
+    # ----------------------------------------------------------------------
+    # Optional hardware/backend-specific metadata
```

Comment:
These properties should at least be typed and they can only return None if that is meaningful. i.e. what does it mean for a backend to have `backend.num_qubits is None` ? And how would the transpiler deal with that?